File: Java/39. Combination Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        helper(candidates, res, list, target, 0);
        return res;
    }
    public void helper(int[] candidates, List<List<Integer>> res, List<Integer> list, int target, int idx){
        if(target < 0){
            return;
        }
        else if(target == 0){
            res.add(new ArrayList<>(list));
        }
        else{
            for(int i=idx; i<candidates.length; i++){
                list.add(candidates[i]);
                helper(candidates,res,list,target-candidates[i],i);
                list.remove(list.size()-1);
            }
        }
    }
}

File: C#/39. Combination Sum.cs
public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        List<IList<int>> res = new();
        List<int> list = new();
        helper(candidates, res, list, target, 0);
        return res;
    }
    public void helper(int[] candidates, List<IList<int>> res, List<int> list, int target, int idx){
        if(target < 0){
            return;
        }
        else if(target == 0){
            res.Add(new List<int>(list));
        }
        else{
            for(int i=idx; i<candidates.Length; i++){
                list.Add(candidates[i]);
                helper(candidates,res,list,target-candidates[i],i);
                list.RemoveAt(list.Count-1);
            }
        }
    }
}
